src/xdsl/rewriter.py

```diff
@@ -1,259 +1,107 @@
-    """
+        block = op.parent
+        block.erase_op(op, safe_erase=safe_erase)
 
-    @abstractmethod
-    def match_and_rewrite(
-            self, op: Operation,
-            new_operands: List[Optional[SSAValue]]) -> Optional[RewriteAction]:
+    @staticmethod
+    def replace_op(op: Operation,
+                   new_ops: Union[Operation, List[Operation]],
+                   new_results: Optional[List[Optional[SSAValue]]] = None,
+                   safe_erase: bool = True):
         """
-        Match an operation, and optionally returns a rewrite to be performed.
-        `op` is the operation to match, and `new_operands` are the potential new values of the operands.
-        `None` values in new_operands are deleted SSA values.
-        This function returns `None` if the pattern did not match, and a rewrite action otherwise.
+        Replace an operation with multiple new ones.
+        If new_results is specified, map the results of the deleted operations with these SSA values.
+        Otherwise, use the results of the last operation added.
+        None elements in new_results are the SSA values to delete.
+        If safe_erase is False, then operations can be deleted even if they are still used.
         """
-        ...
-
-
-@dataclass(eq=False, repr=False)
-class AnonymousRewritePattern(RewritePattern):
-    """
-    A rewrite pattern encoded by an anonymous function.
-    """
-    func: Callable[[Operation, List[Optional[SSAValue]]],
-                   Optional[RewriteAction]]
+        if op.parent is None:
+            raise ValueError("Cannot replace an operation without a parent")
+        block = op.parent
+
+        if not isinstance(new_ops, list):
+            new_ops = [new_ops]
+        if new_results is None:
+            new_results = [] if len(new_ops) == 0 else new_ops[-1].results
+
+        if len(op.results) != len(new_results):
+            raise ValueError(
+                f"Expected {len(op.results)} new results, but got {len(new_results)}"
+            )
+
+        for old_result, new_result in zip(op.results, new_results):
+            if new_result is None:
+                if safe_erase:
+                    old_result.replace_by(ErasedSSAValue(old_result.typ))
+                elif len(old_result.uses) != 0:
+                    raise Exception(
+                        "SSA value was supposed to be destroyed, but still has uses."
+                    )
+            else:
+                old_result.replace_by(new_result)
 
-    def match_and_rewrite(
-            self, op: Operation,
-            new_operands: List[Optional[SSAValue]]) -> Optional[RewriteAction]:
-        return self.func(op, new_operands)
+        op_idx = block.get_operation_index(op)
+        block.erase_op(op_idx, safe_erase=safe_erase)
+        block.insert_op(new_ops, op_idx)
 
+    @staticmethod
+    def inline_block_at_pos(block: Block, target_block: Block, pos: int):
```

Comment:
Should we maybe check that the ops do no use any of the block arguments? 

Assuming I understand this correctly, this does only copy the operations but does not destroy the block, so there might actually be no nice solution to changing the operands of the copied operations without modifying the actual block.

Reply:
Wow I completely overlooked that!
I'm gonna fix this now, thanks!

Reply:
No worries, I would have overlooked like 10 other things you though about ;P

I only found this while fixing the breaking end-to-end test which produces mlir.